content/blog/2023/2023-06-25-archunit.md

```diff
@@ -0,0 +1,598 @@
+- Less dependencies in the codebase make refactoring and splitting up the codebase easier.
+- Respecting naming conventions makes the code easier to read and understand.
+- A clean architecture can facilitate secure code. 
+
+### Example: Data Encapsulation
+Let's discuss in more detail **how clean architecture can improve security**.
+Here's a practical example of how data encapsulation can prevent data exposure
+and how validating dependencies can help us.
+
+Here's a simple REST API that returns employee data:
+
+```java
+public record Employee(long id, String name, boolean active) { }
+```
+
+```java
+public class EmployeeController {
+  @GET()
+  @Path("/employees")
+  public EmployeeResponse getEmployee() {
+    EmployeeService service = new EmployeeService();
+    return service.getEmployee();
+  }
+}
+```
+
+Easy enough. However, let's say, at a later point in time, we add one more attribute to
+our employee entity:
+
+```java
+public record Employee(long id, String name, boolean active, int salary) { }
+```
+What will happen? As our API operates directly on the employee class, we'll expose the newly added
+attribute in the API. That could be the desired behavior in some situations, however, we might also
+expose new attributes involuntarily. The salary of an employee might be confidential and be adding it
+to the record, we expose that information. Therefore, it's usually better to have separate classes 
+for internal use and the API:
+
+```java
+public record EmployeeResponse(long id, String name, boolean active) { }
+```
+
+with a mapping in the service class: 
+
+```java
+public class EmployeeService {
+  public EmployeeResponse getEmployee() {
+    EmployeeDao employeeDao = new EmployeeDao();
+    Employee employee = employeeDao.findEmployee();
+    return new EmployeeResponse(
+      employee.id(),
+      employee.name(),
+      employee.active()
+    );
+  }
+}
+```
+
+which we then use in the controller:
+```java
+public class EmployeeController {
+  @GET()
+  @Path("/employees")
+  public EmployeeResponse getEmployee() {
+    EmployeeService service = new EmployeeService();
+    return service.getEmployee();
+  }
+}
+```
+
+The following image visualizes the difference between the two approaches:
+{{% image alt="Layer dependency" src="images/posts/enforce-architecture-with-arch-unit/arch-unit-layers.png" %}}
+(1) Shows the architecture without and (2) with a service layer. To keep the architecture clean, the API layer
+should only access the service layer and the service layer only access the domain layer. We should avoid direct 
+access from the API to the domain layer.
+
+## Basic ArchUnit Example
+Let's look at how we can use ArchUnit to create a test for the above example. 
+For that, we'll create a project with the following structure: 
+
+{{% image alt="Layer dependency" src="images/posts/enforce-architecture-with-arch-unit/example-project-structure.png" %}}
```

Comment:
For better understanding in this example, it is advisable to place the model classes within the layer packages. This approach would provide the reader with a clearer depiction of the layers.

Reply:
I have updated the code and image.